dex/pair/src/safe_price.rs

```diff
@@ -91,188 +61,103 @@ pub trait SafePriceModule:
-            .set(max_observations_per_record);
-    }
-
-    fn update_safe_state_on_the_fly(&self) {
-        self.update_safe_state(
-            &self.pair_reserve(&self.first_token_id().get()).get(),
-            &self.pair_reserve(&self.second_token_id().get()).get(),
-        );
-    }
-
-    fn update_safe_state(&self, first_token_reserve: &BigUint, second_token_reserve: &BigUint) {
-        let current_block = self.blockchain().get_block_nonce();
-        let mut current_state = self.get_current_state_or_default();
-        let mut future_state = self.get_future_state_or_default();
-
+    fn update_safe_price(&self, first_token_reserve: &BigUint, second_token_reserve: &BigUint) {
         //Skip executing if reserves are 0. This will only happen once, first add_liq after init.
         if first_token_reserve == &0u64 || second_token_reserve == &0u64 {
             return;
         }
 
-        //Skip executing the update more than once per block.
-        if current_state.contains_block(current_block) {
-            return;
-        }
-
-        //Will be executed just once to initialize the current state.
-        if current_state.is_default() {
-            current_state =
-                CumulativeState::new(current_block, first_token_reserve, second_token_reserve);
-        }
-
-        //Will be executed just once to initialize the future state.
-        if self.has_half_max_observations(&current_state) && future_state.is_default() {
-            future_state =
-                CumulativeState::new(current_block, first_token_reserve, second_token_reserve);
-        }
+        let current_round = self.blockchain().get_block_round();
+        let safe_price_info_mapper = self.safe_price_info();
+        require!(!safe_price_info_mapper.is_empty(), ERROR_SAFE_PRICE_INFO);
+        let mut safe_price_info = safe_price_info_mapper.get();
+
+        let pending_price_observation_mapper = self.pending_price_observation();
+        let pending_price_observation: PriceObservation<<Self as ContractBase>::Api> =
+            if !pending_price_observation_mapper.is_empty() {
+                pending_price_observation_mapper.get()
+            } else {
+                PriceObservation::default()
+            };
+        let price_observations_mapper = self.price_observations();
+        let mut price_observations = if !price_observations_mapper.is_empty() {
+            price_observations_mapper.get()
+        } else {
+            ManagedVec::new()
+        };
 
-        //At this point, future state is already initialized and contains half
-        //of the observations that the current state contains.
-        if self.has_max_observations(&current_state) {
-            current_state = future_state.clone();
-            future_state =
-                CumulativeState::new(current_block, first_token_reserve, second_token_reserve);
+        // Save the previously computed price observation, if it's the last one from the previous block
```

Comment:
not like this - this is too complicated. It can be coded in an easier way.

Reply:
Updated.